docs/06-contributors/999-rfcs/2023-05-19-target-specific-resource-implementation.md

```diff
@@ -0,0 +1,118 @@
+---
+title: (Draft) Target Specific Resources Implementation
+description: How should you define a custom resource that is implemented differently on different targets
+---
+
+- **Author(s):**: @ekeren
+- **Submission Date**: {2023-05-19}
+- **Stage**: Draft
+- **Stage Date**: {YYYY-MM-DD}
+
+How should a wing developer define a custom target that is implemented differently on different targets
+
+## Background
+
+Wing is unique by allowing the developer to initiating abstract classes, consider the following code
+```ts (wing)
+bring cloud;
+
+let b = new cloud.Bucket();
+```
+
+At a first glance on this line, one could think that this code will initiate the cloud.Bucket instance.
+But in fact that resource is actually an abstract class and an instance of this class is never initialized.
+The actual class that will be initiated is determined by the target of compilation, the actual implementation of this
+dependency injection mechanism is described in this [PR]([url](https://github.com/winglang/wing/pull/1681)).
+
+Currently, wing doesn't have a way to express different implementations for different targets in the language. 
+The actual `cloud.Bucket` implementation for different targets is in typescript.
+The document propose a solution for defining different implementations for different targets in wing
+
+## Requirements
+
+The proposal should fulfill the following requirements:
+
+- Allow defining target-specific implementations for custom resources.
+- Maintain consistency with the existing language syntax and design.
+- Enable easy integration with the current dependency injection mechanism.
+
+## Proposal
+
+Using familiar abstract class syntax with the `resource_name`.`target_name`.
+
+When an abstract resource gets instantiated, the system looks for a `resource_name.target_name` class that extends that resource
+
+Consider this KVStore.w file (that includes sim and tf-aws impl):
+```ts (wing)
+
+abstract resource KVStore {
+  
+  abstract on_upload(inflight fn: (str, str)): void; 
+  
+  abstract inflight put(k: str, v:str): void;
+  
+  inflight put_json(k: str, j: Json) {
+     return this.put(k, Json.stringify(v));
+  }
+}
+
+resource KVStore.tf-aws extends KVStore {
+  init() { 
+    // Add tf-aws impl
+  } 
+  
+  on_upload(inflight fn: (str, str)): void {
+    // Add tf-aws impl
+  }
+  
+  inflight put(k: str, v:str): void {
+    // Add tf-aws impl
+  }
+} 
```

Comment:
I think a decorator would be a great way to satisfy this usecase:

```js
@Target("tf-aws")
resource TfAwsKvStore extends KvStore {
  // ...
```

I already proposed [using it for `extern`](https://github.com/winglang/wing/issues/478#issuecomment-1465270334), which [Elad didn't like](https://github.com/winglang/wing/issues/478#issuecomment-1465272051). But I think it's a really great fit here.

Reply:
2 cents - so I've come around to the decorators idea, partly after realizing Rust's macros are similar in a way. (For example, in Rust, `#[cfg(test)]` indicates a piece of code should only get compiled when you are testing.) I think an annotation system could be a cheap way to handle these use cases where we want to include extra information for the compiler about resources/fields/functions. 👍 

I only wonder if we should consider different syntaxes, since some developers may associate `@decorator` syntax with how decorators work in Python and TypeScript, where they wrap one function in another function (and I'm not sure if that's the implementation we need to use).

Reply:
@skinny85 , in your example `TfAwsKvStore` is  actually a dummy name that is never used in the code, right?

Reply:
It's never used by the customer code, but I assume Wing’s built-in dependency injection will use it to instantiate the correct class when compiling the code. So, it will be present in the outputted JavaScript.

Reply:
BTW, 
@eladb suggested another syntax option, something like:

```ts (wing)
resource KvStore for "tf-aws"  {
}
```

Reply:
I’m not sure this feature is worth adding new syntax to the language.

Reply:
Basically, I'm worried that there will be more metadata needed to be added to resources in the future, and using something like `for "tf-aws"` won't be able to accommodate those requirements easily.

For example, if you wanted to signify a given resource requires a certain minimum Wing version, with a decorator, it could be something as simple as:

```js
@Target("tf-aws", minVersion = "0.3.4")
```

While with `for "tf-aws"`, I don't see a great option for adding `minVersion`.

Just my 2 cents!